### PR TITLE
replay bug and other code refactoring

### PR DESCRIPTION
**Pull request by rohitpanwar**
https://cdc-dexops.atlassian.net/browse/UPLOAD-1055

### CODE REVIEW

File: upload-processor/BulkFileUploadFunctionApp/Model/BlobCopyRetryEvent.cs [hunk 1]
```diff
@@ -11,8 +12,20 @@ public class BlobCopyRetryEvent
         [JsonPropertyName("retryAttempt")]
         public int retryAttempt { get; set; }
 
-        [JsonPropertyName("sourceBlobUri")]
-        public string? sourceBlobUri { get; set; }
+        [JsonPropertyName("uploadId")]
```
Comment (cfarmer-fearless) [inline upload-processor/BulkFileUploadFunctionApp/Model/BlobCopyRetryEvent.cs:15]: Combine to `CopyPrereq` object

File: upload-processor/BulkFileUploadFunctionApp/Model/CopyPreqs.cs [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+            set { destinationMetadata = value; }
+        }
+
+        public Trace Trace
+        {
+            get { return trace; }
+            set { trace = value; }
+        }
+
+        public CopyPreqs()
```
Comment (cfarmer-fearless) [inline upload-processor/BulkFileUploadFunctionApp/Model/CopyPreqs.cs:87]: Use `.Default`

**Review by cfarmer-fearless — CHANGES_REQUESTED** [review]

File: upload-processor/BulkFileUploadFunctionApp/Services/UploadProcessingService.cs [hunk 3]
```diff
@@ -3,9 +3,10 @@
...
 using Microsoft.Extensions.Configuration;
 using Azure.Identity;
+using System.Text.Json;
+
 using BulkFileUploadFunctionApp.Utils;
 using BulkFileUploadFunctionApp.Model;
-using System.Text.Json;
 
 namespace BulkFileUploadFunctionApp.Services
 {
```
Comment (cfarmer-fearless) [inline upload-processor/BulkFileUploadFunctionApp/Services/UploadProcessingService.cs:12]: Remove unused private variables `_configuration` and `_uploadConfigStorageContainer`

**Review by cfarmer-fearless — CHANGES_REQUESTED** [review]

File: upload-processor/BulkFileUploadFunctionApp/UploadProcessorReplay.cs [hunk 4]
```diff
@@ -5,9 +5,11 @@
 using Microsoft.Extensions.Logging;
```
Comment (cfarmer-fearless) [inline upload-processor/BulkFileUploadFunctionApp/UploadProcessorReplay.cs:5]: Remove unused imports `System.Text.Json` and `Azure.messaging.eventhubs.consumer`

File: upload-processor/BulkFileUploadFunctionApp/UploadProcessorReplay.cs [hunk 5]
```diff
@@ -31,6 +40,12 @@ public UploadProcessorReplay(ILoggerFactory loggerFactory, IUploadEventHubServic
...
             _consumerGroup = Environment.GetEnvironmentVariable("AzureEventHubConsumerGroup", EnvironmentVariableTarget.Process);
+
+            _dexAzureStorageAccountName = Environment.GetEnvironmentVariable("DEX_AZURE_STORAGE_ACCOUNT_NAME", EnvironmentVariableTarget.Process) ?? "";
+            _dexAzureStorageAccountKey = Environment.GetEnvironmentVariable("DEX_AZURE_STORAGE_ACCOUNT_KEY", EnvironmentVariableTarget.Process) ?? "";
+
+            _dexStorageAccountConnectionString = $"DefaultEndpointsProtocol=https;AccountName={_dexAzureStorageAccountName};AccountKey={_dexAzureStorageAccountKey};EndpointSuffix=core.windows.net";
+            _replayCheckpointContainer = "replay-checkpoint";
         }
 
         [Function("UploadProcessorReplay")]
```
Comment (cfarmer-fearless) [inline upload-processor/BulkFileUploadFunctionApp/UploadProcessorReplay.cs:51]: Missing an `await` on line 58 for `ProcessReplayEventHubEventsAsync`

File: upload-processor/BulkFileUploadFunctionApp/UploadProcessorRetry.cs [hunk 6]
```diff
@@ -61,25 +54,27 @@ private async Task ProcessCopyRetryEvent(BlobCopyRetryEvent blobCopyRetryEvent)
         {
             try
             {
-                if(blobCopyRetryEvent.copyRetryStage == null)
+                if(blobCopyRetryEvent.CopyRetryStage == null)
```
Comment (cfarmer-fearless) [inline upload-processor/BulkFileUploadFunctionApp/UploadProcessorRetry.cs:57]: Unnecessary if statement.  The copy stage will never be null.

File: upload-processor/BulkFileUploadFunctionApp/UploadProcessorRetry.cs [hunk 7]
```diff
@@ -61,25 +54,27 @@ private async Task ProcessCopyRetryEvent(BlobCopyRetryEvent blobCopyRetryEvent)
...
-                    switch (blobCopyRetryEvent.copyRetryStage)
+                    switch (blobCopyRetryEvent.CopyRetryStage)
                     {
                         case BlobCopyStage.CopyToDex:
                             try
                             {
-                                await _uploadProcessingService.ProcessBlob(blobCopyRetryEvent.sourceBlobUri);
+                                CopyPrereqs copyPrereqs = await _uploadProcessingService.GetCopyPrereqs(blobCopyRetryEvent.CopyPrereqs.SourceBlobUrl);
+
+                                _uploadProcessingService.CopyAll(copyPrereqs);
```
Comment (cfarmer-fearless) [inline upload-processor/BulkFileUploadFunctionApp/UploadProcessorRetry.cs:77]: Add `await`

**Review by cfarmer-fearless — CHANGES_REQUESTED** [review]

**Comment by cfarmer-fearless** [comment]
Seeing the following copy prereq while testing:
```
Copy preqs: {"UploadId":null,"SourceBlobUrl":"https://ocioededataexchangedev.blob.core.windows.net/bulkuploads/tus-prefix/23434103d60b20c497bbad8012115e86","TusPayloadFilename":"/tus-prefix/23434103d60b20c497bbad8012115e86","DexBlobUrl":null,"DestinationId":"dextesting","EventType":"testevent1","DexBlobFolderName":"dextesting-testevent1","DexBlobFileName":"2024/03/04/10KB-test-file_638451874617914803","Metadata":{"meta_destination_id":"dextesting","meta_ext_event":"testevent1","meta_ext_source":"INTEGRATION-TEST","sender_id":"INTEGRATION-TEST","data_producer_id":"","jurisdiction":"","received_filename":"10KB-test-file","data_stream_id":"dextesting","data_stream_route":"testevent1","version":"2.0","tus_tguid":"23434103d60b20c497bbad8012115e86","upload_id":"23434103d60b20c497bbad8012115e86","trace_id":"TracingDisabled","span_id":"TracingDisabled"},"Targets":[{"target":"dex_edav"},{"target":"dex_routing"}],"Trace":{"TraceId":"TracingDisabled","SpanId":"TracingDisabled","UploadId":"TracingDisabled","DestinationId":null,"EventType":null}}
```
Note that UploadId and DexBlobUrl are null, but there are fields called `upload_id` and `SourceBlobUrl` that are populated.  Should these fields be combined?

File: upload-processor/BulkFileUploadFunctionApp/Services/UploadProcessingService.cs [hunk 8]
```diff
@@ -141,179 +119,156 @@ public async Task ProcessBlob(string blobCreatedUrl)
...
                 // There are some restrictions on container names -- underscores not allowed, must be all lowercase
                 destinationContainerName = $"{destinationId.ToLower()}-{eventType.ToLower()}";
 
-                // Copy the blob to the DeX storage account specific to the program, partitioned by date
-                string dexBlobUrl = await CopyBlobFromTusToDex(uploadedFilePathname, destinationContainerName, destinationBlobFilename, tusInfoFile.MetaData);
-
-                await CopyBlobFromDexToTarget(dexBlobUrl, destinationId, eventType, destinationContainerName, destinationBlobFilename, tusInfoFile.MetaData);                
+                // Get copy targets
+                CopyTarget[] targets = GetCopyTargets(destinationId, eventType);
+                _logger.LogInformation($"Copy Targets: {targets}");
```
Comment (cfarmer-fearless) [inline upload-processor/BulkFileUploadFunctionApp/Services/UploadProcessingService.cs:131]: You'll need to either serialize this to a JSON string or loop through each item in the list (probably using Linq) in order to print out the actual values.  This just prints the class name.

**Review by cfarmer-fearless — CHANGES_REQUESTED** [review]

File: upload-processor/BulkFileUploadFunctionApp/UploadProcessorReplay.cs (hunk 5 above)
Comment (rohitpanwar) [reply upload-processor/BulkFileUploadFunctionApp/UploadProcessorReplay.cs:51]: this is on purpose. Do not want to wait for everything to complete before returning back Http response

File: upload-processor/BulkFileUploadFunctionApp/UploadProcessorRetry.cs (hunk 6 above)
Comment (rohitpanwar) [reply upload-processor/BulkFileUploadFunctionApp/UploadProcessorRetry.cs:57]: will remove this extra check

File: upload-processor/BulkFileUploadFunctionApp/UploadProcessorRetry.cs (hunk 7 above)
Comment (rohitpanwar) [reply upload-processor/BulkFileUploadFunctionApp/UploadProcessorRetry.cs:77]: done

File: upload-processor/BulkFileUploadFunctionApp/Services/UploadProcessingService.cs (hunk 3 above)
Comment (rohitpanwar) [reply upload-processor/BulkFileUploadFunctionApp/Services/UploadProcessingService.cs:12]: done

File: upload-processor/BulkFileUploadFunctionApp/UploadProcessorReplay.cs (hunk 4 above)
Comment (rohitpanwar) [reply upload-processor/BulkFileUploadFunctionApp/UploadProcessorReplay.cs:5]: done

File: upload-processor/BulkFileUploadFunctionApp/Services/UploadProcessingService.cs (hunk 8 above)
Comment (rohitpanwar) [reply upload-processor/BulkFileUploadFunctionApp/Services/UploadProcessingService.cs:131]: removing this as I am printing copy prereqs elsewhere for debugging purposes

**Review by cfarmer-fearless — APPROVED** [review]

**Pull request merged**